File: app/src/main/java/ch/heigvd/pro/b04/android/Poll/PollViewModelFactory.java
package ch.heigvd.pro.b04.android.Poll;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import ch.heigvd.pro.b04.android.Poll.PollViewModel;

/**
 * An implementation of a {@link ViewModelProvider.Factory} that supports the creation of some
 * {@link PollViewModel}s with specific authentication tokens, and dedicated poll identifiers as
 * well.
 */
public class PollViewModelFactory implements ViewModelProvider.Factory {

    private Application application;

    private String token;
    private int idModerator;
    private int idPoll;

    public PollViewModelFactory(
            @NonNull Application application,
            String token,
            int idModerator,
            int idPoll
    ) {
        this.application = application;
        this.token = token;
        this.idModerator = idModerator;
        this.idPoll = idPoll;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new PollViewModel(application, token, idModerator, idPoll);
    }
}

File: app/src/main/java/ch/heigvd/pro/b04/android/Poll/PollViewModelFactory.kt
package ch.heigvd.pro.b04.android.Poll

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class PollViewModelFactory(
    private val application: Application,
    private val idModerator: Int,
    private val idPoll: Int,
    private val token: String
) : ViewModelProvider.Factory {

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        val constructors = modelClass.declaredConstructors
        return constructors[0].newInstance(application, idModerator, idPoll, token) as T
    }
}
